rust/experimental/lang/src/main.rs

```diff
@@ -14,142 +14,277 @@
 
 //! A language syntax coloring and indentation plugin for xi-editor.
 
-#[macro_use]
+extern crate xi_core_lib;
 extern crate xi_plugin_lib;
+extern crate xi_rope;
+extern crate xi_trace;
 
-use xi_plugin_lib::caching_plugin::{self, PluginCtx, SpansBuilder};
+use std::{collections::HashMap, env, path::Path};
 
-use std::env;
+use language::{plaintext::PlaintextParser, rust::RustParser};
+use parser::Parser;
+use statestack::{HolderNewState, State};
+use xi_core_lib::{plugins::rpc::ScopeSpan, ConfigTable, LanguageId, ViewId};
+use xi_plugin_lib::{mainloop, Cache, Plugin, StateCache, View};
+use xi_rope::RopeDelta;
+use xi_trace::{trace, trace_block, trace_payload};
 
-mod rust;
-mod colorize;
-mod statestack;
+mod language;
+mod parser;
 mod peg;
+mod statestack;
+
+const LINES_PER_RPC: usize = 50;
+
+type Scope = Vec<String>;
```

Comment:
this is less useful now that we're using integer id's. I don't think it's necessary now, but if we were polishing this i might suggest `ScopeId` typedef.